md5sum_cli.cpp

```diff
@@ -0,0 +1,268 @@
+
+    // With --check, don't generate any output.
+    // The exit code indicates success or failure.
+    bool status_only = false;
+
+    // With --check, suppress the "OK" printed for each verified file.
+    bool quiet = false;
+
+    GetOpt getopt_long
+        (argc
+        ,argv
+        ,"chv"
+        ,long_options
+        ,&option_index
+        ,1
+        );
+
+    while(EOF != (c = getopt_long ()))
+        {
+        switch (c)
+            {
+            case 'b':
+                binary = true;
+                have_input_mode_option = true;
+                break;
+            case 'c':
+                do_check = true;
+                break;
+            case QUIET_OPTION:
+                quiet = true;
+                break;
+            case STATUS_OPTION:
+                status_only = true;
+                break;
+            case 't':
+                have_input_mode_option = true;
+                break;
+            case HELP_OPTION:
+                show_help = true;
+                break;
+            case VERSION_OPTION:
+                show_version = true;
+                break;
+            default:
+                // Error message was already given from getopt() code, so no need
+                // to output anything else here, but do flush its output so that it
+                // appears before the usage message.
+                std::fflush(stderr);
+
+                command_line_syntax_error = true;
+            }
+
+        if(command_line_syntax_error)
+            break;
+        }
+
+    if(command_line_syntax_error)
+        {
+        std::cerr << "Try 'md5sum --help' for more information." << std::endl;
+        return EXIT_FAILURE;
+        }
+    if(have_input_mode_option && do_check)
+        {
+        std::cerr
+            << "The --binary and --text options are meaningless when "
+            << "verifying checksums." << std::endl;
+        return usage(EXIT_FAILURE);
+        }
+    if(status_only && !do_check)
+        {
+        std::cerr
+            << "The --status option is meaningful only when verifying checksums."
+            << std::endl;
+        return usage(EXIT_FAILURE);
+        }
+    if(quiet && !do_check)
+        {
+        std::cerr
+            << "The --quiet option is meaningful only when verifying checksums."
+            << std::endl;
+        return usage(EXIT_FAILURE);
+        }
+    if(show_help)
+        {
+        return usage(EXIT_SUCCESS);
+        }
+
+    if(show_version)
+        {
+        version();
+        return EXIT_SUCCESS;
+        }
+
+    for(int i = getopt_long.optind; i < argc; ++i)
+        {
+        char const* filename = argv[i];
+        std::string error_message;
+        std::string md5;
+
+        if(do_check)
+            {
+            std::vector<md5sum_for_file> sums;
+
+            try
+                {
+                sums = md5sum_vec_from_file(filename);
+                }
+            catch(std::runtime_error const& e)
```

Comment:
Why do we duplicate this catch block thrice here instead of putting it around all this code? If there is no special reason to do it, it would be definitely better to have it in only one place.